nengo_ocl/clra_gemv.py

```diff
@@ -1238,43 +1240,376 @@ def choose_plans(self):
+    hA : np.array or scipy.sparse.spmat
+        Matrix to multiply in any Numpy or Scipy format. ``Y = hA.dot(X)``
     X, Y : CLRaggedArrays of length 1
         Input/output data.
     inc : bool
-        Whether to increment ``Y`` (True), or set it (False).
+        Whether the plan will increment or set ``Y``.
+    tag : str
+        Tag for the plan.
+    """
+
+    def __init__(
+        self,
+        queue,
+        hA,
+        X,
+        Y,
+        inc=False,
+        tag=None,
+    ):
+        """Initializing will
+            - format the sparse matrix
+            - transfer matrix to device
+            - check the validity of device data
+            - create the plans
+        Most often, the caller will use prog_obj.plans after initializing
+        """
+        self.queue = queue
+        self.hA = hA  # Host matrix in numpy or scipy sparse with methods such as "dot"
+        self.X = X
+        self.Y = Y
+        self.inc = inc
+        self.tag = tag
+
+        self.hAstruct = None  # Host array data. Depends on storage format.
+        self.dAstruct = None  # Device array data. Mirror of hAstruct.
 
-    Notes
-    -----
-    This function crashes when there are >10M nonzero weights. A potential solution
-    would be some way to tell each work item to do multiple rows.
+        self.to_hostdata()
+        self.to_device()
+        self.validate_data()
+        self.plans = self.choose_plans()
+        self.assign_plan_characteristics(self.plans)
+
+    def assign_plan_characteristics(self, plans):
+        pass
+
+    def choose_plans(self):
+        raise NotImplementedError("Subclass must implement")
+
+    def to_device(self):
+        raise NotImplementedError("Subclass must implement")
+
+    def to_hostdata(self):
+        raise NotImplementedError("Subclass must implement")
+
+    def validate_data(self):
+        raise NotImplementedError("Subclass must implement")
+
+
+class plan_ellpack(spmv_prog):
+    """SPMV program with the ELLPACK format.
+
+    This class chooses the implementation based on the size of the problem.
     """
-    assert len(X) == len(Y) == 1
 
-    for arr in [X, Y]:
-        assert (arr.stride1s == 1).all()
-        if not ((arr.shape1s == 1).all() and (arr.stride0s == 1).all()):
-            raise NotImplementedError(
-                "OCL SparseDot only supports matrix-vector currently, not matrix-matrix"
+    @staticmethod
+    def scipy2elldata(scipy_mat, force_nonempty=True):
+        """ Works with scipy sparse or numpy dense argument """
+        scipy_mat = scipy_mat.copy()
+        try:
+            lilmat = scipy_mat.tolil()
+        except AttributeError:
+            lilmat = scipy_sparse.lil_matrix(scipy_mat)
+
+        shape = scipy_mat.shape
+        rowlens = np.array([len(rowdata) for rowdata in lilmat.rows])
+        ellwidth = max(rowlens)
+        ind_dtype = (
+            np.int32 if len(rowlens) < 2 ** 32 and ellwidth < 2 ** 32 else np.int64
```

Comment:
Since that'd be a huge matrix, I'm guessing this hasn't been tested (also, it should technically be `2**31`, since that's the max value for signed `int32`). I think we should just hard-code this as `int32`, since I don't see us hitting that limit any time soon (someone would need to have more than 2 billion neurons in an ensemble), and probably there would be further considerations if we're moving into that size of matrix.

Reply:
Good point!